accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -992,9 +996,12 @@ arguments) marked as legacy types.
 
 If a class `C` in a legacy library implements the same generic class `I` more
 than once, it is an error if the `LEGACY_ERASURE` of all such super-interfaces
-are not all syntactically equal.  For the purposes of runtime subtyping checks,
-`C` is considered to implement the canonical `LEGACY_ERASURE` of the
-super-interfaces in question.
+are not all syntactically equal.
+
+When `C` implements `I` once, and also when `C` implements `I` more than once
```

Comment:
So just to be clear, I think this means that if a legacy class `B` has `A<int?>` as a super-interface, then in an opted in library, an assignment of the form `A<int> x = B()` is not considered an error?

Reply:
Right.

This is a loss in static type safety. As I mentioned, it was motivated by the rule that a legacy class should only have legacy properties.

We can of course offer additional help to developers in their quest to detect discrepancies like this (for instance, a lint or a hint could flag such situations as dangerous).

However, `B` could change the actual member signatures and I do not believe that we'd have a rule justifying that `B` would then implement `I` in a different way:

```dart
// Opted in.
abstract class A<X> {
  X get x;
}

abstract class A1 extends A<int> {}

// Opted out.
class B implements A1 {
  int get x => null;
}

// Opted in.
class C implements B {}

void main() {
  A<int> a = C(); // OK!
}
```

So here we trust `C` to implement `A<int>`, even though the overriding implementation of `x` in `B` implies that `x` may actually return null (which means that it is reasonable to say that `C` implements `A<int*>`, but misleading to say that it implements `A<int>`).

So even though it looks like we are preserving more static type information, I suspect that it may not be sufficiently reliable to be very helpful.